src/cdcacm.c

```diff
@@ -0,0 +1,364 @@
+        (*buf)[1] = 100; // ms
+        (*buf)[2] = 0;
+        (*buf)[3] = 0;
+        (*buf)[4] = DFU_STATUS_OK;
+        (*buf)[5] = 0;
+        return USBD_REQ_HANDLED;
+    case DFU_DETACH:
+        re_enter_bootloader = true;
+        return USBD_REQ_HANDLED;
+    }
+    return USBD_REQ_NOTSUPP;
+}
+
+#define USB_MSG_MAXLEN 64
+char usb_msg_buffer[USB_MSG_MAXLEN];
+int usb_msg_len = 0;
+
+static void cdcacm_data_rx_cb(usbd_device *usbd_dev, uint8_t ep) {
+    (void)ep;
+
+    int remaining_len = (USB_MSG_MAXLEN - 1) - usb_msg_len;
+    char* new_data_ptr = usb_msg_buffer + usb_msg_len;
+
+    int len = usbd_ep_read_packet(usbd_dev, 0x01, new_data_ptr, remaining_len);
+
+    if (len) {
+        usb_msg_len += len;
+        usb_msg_buffer[usb_msg_len] = '\0'; // add null terminator to make it a string
+
+        char* end_of_msg = strchr(usb_msg_buffer, '\n');  // test if \n in buffer
+        char response_buffer[64];
+        char* response_ptr = response_buffer;
+        int full_response_len = 0;
+
+        while (end_of_msg != NULL) {
+            *end_of_msg = '\0';  // replace newline with null terminator
+            char* carriage_return = strchr(usb_msg_buffer, '\r');
+            if (carriage_return) {
+                *carriage_return = '\0';  // remove a \r
+            }
+
+            int msg_len = end_of_msg - usb_msg_buffer + 1;
+
+            int usb_response_len = parse_msg(usb_msg_buffer, response_ptr, 63 - full_response_len);
+            response_ptr[usb_response_len++] = '\n';  // replace null-terminator with newline
+            full_response_len += usb_response_len;
+            response_ptr += usb_response_len;
+
+            usb_msg_len -= msg_len;
+            if (usb_msg_len < 0) {
+                usb_msg_len = 0;
+            }
+            if (usb_msg_len > 0) {
+                // move remaining data to start of buffer
+                memmove(usb_msg_buffer, end_of_msg + 1, usb_msg_len);
+            }
+
+            // repeat if \n in buffer
+            end_of_msg = strchr(usb_msg_buffer, '\n');  // test if \n in buffer
+        }
+
+        // drop a full buffer without newlines
+        if (usb_msg_len == (USB_MSG_MAXLEN - 1)) {
+            usb_msg_len = 0;
+        }
+
+        if (full_response_len > 0) {
+            usbd_ep_write_packet(usbd_dev, 0x82, response_buffer, full_response_len);
+        }
+    }
+}
+
+static void cdcacm_set_config(usbd_device *usbd_dev, uint16_t wValue) {
+    (void)wValue;
+
+    usbd_ep_setup(usbd_dev, 0x01, USB_ENDPOINT_ATTR_BULK, 64, cdcacm_data_rx_cb);
+    usbd_ep_setup(usbd_dev, 0x82, USB_ENDPOINT_ATTR_BULK, 64, NULL);
+    usbd_ep_setup(usbd_dev, 0x83, USB_ENDPOINT_ATTR_INTERRUPT, 16, NULL);
+
+    usbd_register_control_callback(
+                usbd_dev,
+                USB_REQ_TYPE_CLASS | USB_REQ_TYPE_INTERFACE,
+                USB_REQ_TYPE_TYPE | USB_REQ_TYPE_RECIPIENT,
+                cdcacm_control_request);
+
+    // Indicate we've enumerated
```

Comment:
Is this the same behaviour as before? We may need to update docs and communicate to volunteers.

Reply:
This behaviour is unchanged.